certbot/certbot/_internal/client.py

```diff
@@ -454,7 +454,22 @@ def obtain_certificate(self, domains: List[str], old_keypath: Optional[str] = No
                 os.remove(csr.file)
             return self.obtain_certificate(successful_domains)
         else:
-            cert, chain = self.obtain_certificate_from_csr(csr, orderr)
+            try:
+                cert, chain = self.obtain_certificate_from_csr(csr, orderr)
+            except messages.Error as error:
+                if self.config.allow_subset_of_names and error.subproblems is not None:
+                    failed_domains = [problem.identifier.value for problem in error.subproblems
+                                        if problem.identifier is not None]
+                    successful_domains = [x for x in domains if x not in failed_domains]
+                    if successful_domains != domains:
+                        logger.info("error obtaining certificate with --allow-subset-of-names "
+                            "specified: retrying without %s", failed_domains)
+                        if not self.config.dry_run:
+                            os.remove(key.file)
+                            os.remove(csr.file)
+                        cert, chain, key, csr = self.obtain_certificate(successful_domains)
```

Comment:
Check that `successful_domains` is not empty. Otherwise:

<img width="1041" alt="image" src="https://user-images.githubusercontent.com/311534/168507746-06421e0a-6e0c-4971-b8c6-12f6e501d060.png">

Reply:
Check added - just re-raising the original error if there are no retryable domains.
Also added tests for this scenario during order creation and finalization.